connection.go

```diff
@@ -497,6 +541,29 @@ func (c *Connection) Reply(ctx context.Context, opErr error) {
 	}
 }
 
+// The NotifyKernel is same as Reply func of Connection.But the diff is
+// that the func only send to kernel.
+func (c *Connection) NotifyKernel(ctx context.Context) {
+
+	// we should get outmsg from context
+	var key interface{} = contextKey
+	foo := ctx.Value(key)
```

Comment:
Can you explain why we need to pull this value out of the context instead of just passing it as a parameter to NotifyKernel?

Reply:
Just do like the reply func.  Of course you can also just pass the value of the context as a parameter to NotifyKernel.

Reply:
Thanks. The reply mechanism needs to use a context because it passes a value (https://github.com/jacobsa/fuse/blob/659cc51b20ee0bf7bf68875df8ea1a0e0704ba42/connection.go#L406) across API boundaries (to the file system implementation and back).

This is not the case here AIUI, so please change the function signature to not take a context, but just take the opState parameter directly.

Reply:
I rethink this case, even the notify func is  diff with the normal op which need reply function. But the context use in notify func  is same as the reply func.   Just the context is created by our func (s *fileSystemServer) Invalidateentry/inode.   And now we call unity all notify op and the normal op.  
For the correct context use, we should alway use the context for notify case. see func (s *fileSystemServer) InvalidateInode for opsInFlight.Done.  

Maybe i am not understandering the issue on detail.

Reply:
> I rethink this case, even the notify func is diff with the normal op which need reply function. But the context use in notify func is same as the reply func. Just the context is created by our func (s *fileSystemServer) Invalidateentry/inode. And now we call unity all notify op and the normal op.

If I understand correctly, you’re saying having both functions (NotifyKernel and Reply) take arguments via a context is good because they are similar, yes?

While similarity is good, I would argue that in this case, it’s clearer if a context is not used unless absolutely necessary. Since we’re creating a context but then immediately call the function (instead of passing the context somewhere else), I don’t think we need to use a context.

Contexts are used (aside from deadlines and cancellation) to propagate data across API boundaries. In this case, we don’t need to cross API boundaries, so we shouldn’t use a context if we can use a simple variable instead.